lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,398 @@
+  /// to avoid invalid values on non-Posix machines.
+  (String, String?) _packageNameAndSubpath(String specifier) {
+    var parts = p.posix.split(specifier);
+    var name = parts.removeAt(0);
+
+    if (name.startsWith('.')) {
+      throw "pkg: name $name must not start with a '.'.";
+    }
+    if (name.contains('\\')) {
+      throw "pkg: name $name must not contain a '\\'.";
+    }
+    if (name.contains('%')) {
+      throw "pkg: name $name must not contain a '%'.";
+    }
+
+    if (name.startsWith('@')) {
+      if (parts.isEmpty) {
+        throw "pkg: name $name is an invalid package name."
+            "Scoped packages, which start with '@', must have a second segment.";
+      }
+      name = p.posix.join(name, parts.removeAt(0));
+    }
+
+    return (name, parts.isNotEmpty ? p.posix.joinAll(parts) : null);
+  }
+
+  /// Returns an absolute path to the root directory for the most proximate
+  /// installed `packageName`.
+  ///
+  /// Implementation of `PACKAGE_RESOLVE` from the [Resolution Algorithm
+  /// Specification](https://nodejs.org/api/esm.html#resolution-algorithm-specification).
+  String? _resolvePackageRoot(String packageName, String basePath) {
+    var baseDirectory = p.dirname(basePath);
+
+    String? recurseUpFrom(String entry) {
+      var potentialPackage = p.join(entry, 'node_modules', packageName);
+
+      if (dirExists(potentialPackage)) return potentialPackage;
+      var parent = p.dirname(entry);
+
+      // prevent infinite recursion
+      if (entry == parent) return null;
+
+      var rootLength = isWindows ? 1 : 0;
+
+      if (Uri.directory(parent).pathSegments.length == rootLength) return null;
+
+      return recurseUpFrom(parent);
+    }
+
+    return recurseUpFrom(baseDirectory);
+  }
+
+  /// Returns a file path specified by the `sass` or `style` values in a package
+  /// manifest, or an `index` file relative to the package root.
+  String? _resolvePackageRootValues(
+      String packageRoot, Map<String, dynamic> packageManifest) {
+    if (packageManifest['sass'] case String sassValue) {
+      if (validExtensions.contains(p.extension(sassValue))) {
+        return p.url.join(packageRoot, sassValue);
+      }
+    }
+
+    if (packageManifest['style'] case String styleValue) {
+      if (validExtensions.contains(p.extension(styleValue))) {
+        return p.url.join(packageRoot, styleValue);
+      }
+    }
+
+    var result = resolveImportPath(p.url.join(packageRoot, 'index'));
+    if (result != null) return result;
+    return null;
+  }
+
+  /// Returns a file path specified by a `subpath` in the `exports` section of
+  /// package.json.
+  ///
+  /// `packageName` is used for error reporting.
+  String? _resolvePackageExports(String packageRoot, String? subpath,
+      Map<String, dynamic> packageManifest, String packageName) {
+    if (packageManifest['exports'] == null) return null;
+    var exports = packageManifest['exports'] as Object;
```

Comment:
```suggestion
    var exports = packageManifest['exports'] as Object?;
    if (exports == null) return null;
```

Not only does this avoid indexing the manifest twice, it's also clearer that the cast is just to convert `dynamic` into something that has actual type checks.